## Overview
The R function **SVreconDELLY** filters somatic structural variant (SV) calls predicted by **DELLY** by attempting to reconstruct their predicted **CONSENSUS/CONTIG** sequence from reference genome sequences.  
If successful, it outputs genomic positions and sequences suitable for downstream **BLAST** and **mutspy** analysis (as described in Ellegaard et al.).

The filtering is strict and strongly favors simple SVs where the predicted SV contig (`CONSENSUS` in the VCF `INFO` field) can be fully reconstructed using only the VCF information and GRCh38 reference genome.

---

## Input Arguments

### 1. `vcf_file`
Path to a `.vcf` or `.vcf.gz` file generated by **DELLY** (PMID: 22962449).  
Requirements:
- Must contain **somatic** SV calls (`SOMATIC` in `INFO`)  
- Only SVs with a predicted **CONSENSUS** sequence in the `INFO` field are processed  
- Pipeline built using `SVMETHOD=EMBL.DELLYv0.9.1`  

### 2. `output_folder`
Directory where output files will be written.  
Three output files per sample are generated (see below).

---

## Running SVreconDELLY
```r
source("path/to/SVreconDELLY_v92.R")
SVreconDELLY("path/to/your/file.vcf", "your/output/folder/")
```

---

## What SVreconDELLY Does
1. Processes only SVs with a `CONSENSUS` sequence in the `INFO` field and removes SVs marked as `IMPRECISE`.  Retains only SVs with `FILTER == PASS`.
2. Optionally filters SVs by `MinMAPQ` and `srMAPQ` in `INFO` field (default `MinMAPQ = 0`).  
3. Reconstructs the predicted SV using breakpoints from the VCF and GRCh38 reference genome.  
4. Inserts predicted breakpoint positions and homology regions (HOMSEQ) into the SV contig (`CONSENSUS` in `INFO`).  
   - Extracts breakpoint‑flanking sequences for mutspy → `SampleID_DELLYrecon_SVmutspy.txt`  
5. Generates genomic donor‑site positions for WGS coverage analysis → `SampleID_DELLYrecon_DONOR.bed`  
6. Produces 50‑bp breakpoint‑centered SV contigs (with predicted homology) for downstream filtering, e.g. by pangenome BLAST as in Ellegaard et al. → `SampleID_DELLYrecon_SV50.fa`  

---

## Output Files
Using the VCF filename (without extension) as `SampleID`, the following files are produced:

1. **`SampleID_DELLYrecon_SVmutspy.txt`**  
   Contains information from SV reconstruction analysis, including SV motifs for mutspy.

2. **`SampleID_DELLYrecon_DONOR.bed`**  
   BED file with donor genomic positions for WGS coverage analysis.

3. **`SampleID_DELLYrecon_SV50.fa`**  
   FASTA file with 50‑bp breakpoint‑centered SV contigs for BLAST searches against reference genomes.

---

## Columns in `SampleID_DELLYrecon_SVmutspy.txt`

| Column | Name | Description |
|-------|------|-------------|
| 1 | `SV_predictions_tool` | Tool used to generate the input VCF (here: `"DELLY"`) |
| 2 | `SVrecon_pipeline` | SVrecon pipeline version |
| 3 | `SVrecon_Parameters` | Values for `GenomicFlank`, `CountFlank`, `maxHom`, `MinMAPQ` |
| 4 | `Sample_ID` | Sample ID derived from VCF filename |
| 5 | `SV_ID` | SV ID extracted from the VCF `ID` column |
| 6 | `Genomic_Position1` | Donor site 1 (`CHROM` + `POS`) |
| 7 | `Genomic_Position2` | Donor site 2 (from `INFO`) |
| 8 | `SVtype` | SV type from VCF `INFO` |
| 9 | `SVtypeClass` | SV type class assigned by SVreconDELLY |
| 10 | `SVLEN` | SV length from VCF `INFO` |
| 11 | `SV_length_estimate` | Approximate SV length (`Position2 - Position1`) |
| 12 | `SV_CONTIG` | SV consensus sequence from VCF `INFO` |
| 13 | `SV_CONTIG_w.Pred.BreakSite` | SV contig with predicted breakpoint `[ ]` and predicted homology regions wihin `[ ]` |
| 14 | `Pred.Breakpoint_Quality` | Quality category of predicted breakpoint within the SV contig: `"Good"`, `"Fair"`, `"Poor"` |
| 15 | `SV_mutspy_input` | SV motif for mutspy analysis (as described in Ellegaard et al.) |

---

# About SVreconDELLY
**Version:** v0.92  
**Author:** Jesper Bramsen (bramsen@clin.au.dk)  
**Date:** 2025‑05‑01  
**Last Update:** 2025‑05‑01  

**Dependencies in R:**  
`dplyr`, `stringr`, `data.table`, `GenomicRanges`, `BSgenome.Hsapiens.UCSC.hg38`, `tidyr`, `Biostrings`

**Built and tested with:**  
R v4.4.2; dplyr v1.1.4; stringr v1.5.1; data.table v1.16.2; GenomicRanges v1.56.2; BSgenome.Hsapiens.UCSC.hg38 v1.4.5; tidyr v1.3.1; Biostrings v2.72.1

---


